### PR TITLE
Fixes #4168 - Add support for Mesos HTTP/TCP health checks (#4210)

### LINKED ISSUE #4168: Support Mesos HTTP/TCP health checks
The Mesos team is working hard on adding support for HTTP/TCP health checks. Their solution will initially reach parity with the currently supported `COMMAND` health checks: [MESOS-5961](https://issues.apache.org/jira/browse/MESOS-5961), [MESOS-3567](https://issues.apache.org/jira/browse/MESOS-3567), [MESOS-2533]( https://issues.apache.org/jira/browse/MESOS-2533).

In order to scale better, Marathon should support these new kinds of health checks as soon as possible.

The idea is to update the app definition, introducing these new health check protocols: `MESOS_HTTP`, `MESOS_HTTPS`, `MESOS_TCP`. They'll behave exactly like the currently supported `COMMAND` health check.

The following needs to be done in order to achieve this:

- [ ] Refactor `HealthCheck`, adding `HttpConf`,`TcpConf`, `MesosHttpConf`, etc. and moving the settings from the main class to these per-protocol config classes.
- [ ] Add JSON formats and validations. Mesos will initially support at-most one health check.
- [ ] Update TaskBuilder logic to account for the new health check types. Marathon will have to implement logic to put the right port number in the health checks proto, taking in consideration the networking specified in the app definition.
- [ ] Integration tests.

This is also related to #3957 (Marathon will currently forget the app health information if an app is using Mesos health checks).

### PR DESCRIPTION
**Pull request by janisz**
This is a cherrypick of a @gkleiman's PR #4210. 
Marathon 1.4 introduced Mesos HTTP/TCP healthchecks. This could result in better performance of a Marathon. Unfortunately Marathon 1.4 introduced a performance degradation making it unusable in bigger installations ([MARATHON-7358](https://jira.mesosphere.com/browse/MARATHON-7358)). This PR add Mesos Healthcecks support to Marathon 1.3. This change is future proof and upgrading to 1.4 is safe.

Moving all Marathon applications from Marathon healthcecks to Mesos Healthcecks can't be done at once. Every application need to be redeployed with new configuration. This PR allow users to slowly move to Mesos Healthchecks and then upgrade to Marathon 1.4

<details>
Manually fixed Conflicts:

-	[`project/build.scala`](https://github.com/janisz/marathon/blob/431de4e8932b048433405131a8875904895f4d73/project/build.scala)
-	[`src/main/java/mesosphere/marathon/Protos.java`](https://github.com/janisz/marathon/blob/431de4e8932b048433405131a8875904895f4d73/src/main/java/mesosphere/marathon/Protos.java)
-	[`src/main/scala/mesosphere/marathon/core/health/HealthCheck.scala`](https://github.com/janisz/marathon/blob/431de4e8932b048433405131a8875904895f4d73/src/main/scala/mesosphere/marathon/core/health/HealthCheck.scala)
-	[`src/main/scala/mesosphere/marathon/core/health/impl/HealthCheckActor.scala`](https://github.com/janisz/marathon/blob/431de4e8932b048433405131a8875904895f4d73/src/main/scala/mesosphere/marathon/core/health/impl/HealthCheckActor.scala)
-	[`src/main/scala/mesosphere/marathon/core/health/impl/MarathonHealthCheckManager.scala`](https://github.com/janisz/marathon/blob/431de4e8932b048433405131a8875904895f4d73/src/main/scala/mesosphere/marathon/core/health/impl/MarathonHealthCheckManager.scala)
-	[`src/main/scala/mesosphere/marathon/state/AppDefinition.scala`](https://github.com/janisz/marathon/blob/431de4e8932b048433405131a8875904895f4d73/src/main/scala/mesosphere/marathon/state/AppDefinition.scala)
-	[`src/test/scala/mesosphere/marathon/api/v2/json/AppDefinitionTest.scala`](https://github.com/janisz/marathon/blob/431de4e8932b048433405131a8875904895f4d73/src/test/scala/mesosphere/marathon/api/v2/json/AppDefinitionTest.scala)
-	[`src/test/scala/mesosphere/marathon/core/health/impl/HealthCheckActorTest.scala`](https://github.com/janisz/marathon/blob/431de4e8932b048433405131a8875904895f4d73/src/test/scala/mesosphere/marathon/core/health/impl/HealthCheckActorTest.scala)
-	[`src/test/scala/mesosphere/marathon/core/health/impl/MarathonHealthCheckManagerTest.scala`](https://github.com/janisz/marathon/blob/431de4e8932b048433405131a8875904895f4d73/src/test/scala/mesosphere/marathon/core/health/impl/MarathonHealthCheckManagerTest.scala)
-	[`src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala`](https://github.com/janisz/marathon/blob/431de4e8932b048433405131a8875904895f4d73/src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala)
-	[`src/test/scala/mesosphere/marathon/state/MarathonStoreTest.scala`](https://github.com/janisz/marathon/blob/431de4e8932b048433405131a8875904895f4d73/src/test/scala/mesosphere/marathon/state/MarathonStoreTest.scala)
</details>

### CODE REVIEW

**Comment by janisz** [comment]
@gkleiman Can you take a look at this?

**Comment by timcharper** [comment]
Thank you, @janisz ! I'll take a look at this Tuesday. (Monday is a holiday) :)

**Comment by timcharper** [comment]
@janisz reviewed; looks like a straight forward (and... probably painful) back-port of the feature. Great work!

I don't see any reason to not just merge this; the enum values are the same Going to request a second opinion from @jdef just to be sure there's not any unexpected consequences of merging.

Have you already deployed this patch to your environment? Do both types of health-checks work as expected?

**Comment by janisz** [comment]
I've only tested it on development cluster. It works as exepcted with no problems. We will push this version to all environments and then slowly introduce MESOS_HTTP healthchecks to newly deployed/updated applications.

**Comment by janisz** [comment]
@timcharper It works at our infrastructure for 3 weeks. We are slowly moving applications to Mesos Native healthchecks. Now about 35% of our apps has MESOS_HTTP healthcheck on our biggest cluster. During this process we didn't observed any issues.

I think it's ready to merge.

**Comment by janisz** [comment]
Below is a graph presenting Marathon requests time for 1.3.10/1.4.2 and 1.3.10 with this patch and enabled Mesos Healthchecks. Currently only 40% of healthchecks are done by Mesos. As you can see responses times are more stable. It's working for a while for us. Other metrics also looks more stable then with Marathon Healthcekcs.

We roll backed from 1.4.2 because it has performance issues.

marathon-consul-test2-mesos-http

**Comment by janisz** [comment]
@unterstein I added commit with AppDefinition validator to explicitly enable Mesos HealthChecks before use.

File: docs/docs/command-line-flags.md [hunk 1]
```diff
@@ -47,6 +47,7 @@ The core functionality flags can be also set by environment variable `MARATHON_O
     - "vips" can be used to enable the networking VIP integration UI.
     - "task\_killing" can be used to enable the TASK\_KILLING state in Mesos (0.28 or later)
     - "external\_volumes" can be used if the cluster is configured to use external volumes.
+    - "mesos_healthchecks" can be used if ti enable the Mesos HealthChecks (Mesos 1.1 or later).
```
Comment (unterstein) [inline docs/docs/command-line-flags.md]: I think there is some black magic rule, why the others are using `\_`

File: src/main/scala/mesosphere/marathon/Features.scala [hunk 2]
```diff
@@ -17,12 +17,16 @@ object Features {
...
+  lazy val MESOS_HEALTHCHECKS = "mesos_healthchecks"
+
   lazy val availableFeatures = Map(
     VIPS -> "Enable networking VIPs UI",
     TASK_KILLING -> "Enable the optional TASK_KILLING state, available in Mesos 0.28 and later",
     EXTERNAL_VOLUMES -> "Enable external volumes support in Marathon",
     SECRETS -> "Enable support for secrets in Marathon (experimental)",
-    GPU_RESOURCES -> "Enable support for GPU in Marathon (experimental)"
+    GPU_RESOURCES -> "Enable support for GPU in Marathon (experimental)",
+    MESOS_HEALTHCHECKS -> "Enable support for Mesos Native Healthchecks in Marathon (require Mesos version >= 1.1.0)"
```
Comment (unterstein) [inline src/main/scala/mesosphere/marathon/Features.scala]: I think mesos health checks were introduced in `1.2.0` according to https://mesosphere.com/blog/2017/01/05/introducing-mesos-native-health-checks-apache-mesos-part-1/ ?

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala [hunk 3]
```diff
@@ -769,6 +770,13 @@ object AppDefinition extends GeneralPurposeCombinators {
     }
   }
 
+  private val complyWithMesosHealthChecks: Validator[AppDefinition] =
+    conditional[AppDefinition](appDef =>
+      (appDef.healthChecks.count(_.isInstanceOf[MesosHealthCheck]) -
+        appDef.healthChecks.count(_.isInstanceOf[MesosCommandHealthCheck])) > 0) {
+      featureEnabled(Features.MESOS_HEALTHCHECKS)
```
Comment (unterstein) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: I would add something like this additionally:

```
val mesosHealthCheckMinVersion = SemanticVersion(1, 2, 0)
...
featureEnabled(Features.MESOS_HEALTHCHECKS) && LibMesos.version >= mesosHealthCheckMinVersion
```

What do you think?

File: docs/docs/command-line-flags.md (hunk 1 above)
Comment (unterstein) [inline docs/docs/command-line-flags.md]: I think mesos health checks were introduced in `1.2.0` according to https://mesosphere.com/blog/2017/01/05/introducing-mesos-native-health-checks-apache-mesos-part-1/ ?

Comment (janisz) [inline docs/docs/command-line-flags.md]: I don't know why it's done that way. Probably to prevent changing font _style_. When you take a look at [rendered html it looks the same](https://github.com/janisz/marathon/blob/ed340f969186ae99d43ad10cc42aadeae2997980/docs/docs/command-line-flags.md#optional-flags).
IMO this should be wrapped in code format (\`) not double quotes (`"`)

File: src/main/scala/mesosphere/marathon/Features.scala (hunk 2 above)
Comment (janisz) [inline src/main/scala/mesosphere/marathon/Features.scala]: This blog post is partially correct. Mesos Health Checks were introduced as an **experimental** feature in [Mesos 1.1.0](https://github.com/apache/mesos/blob/1.1.0/CHANGELOG#L15-L30)

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala (hunk 3 above)
Comment (janisz) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: Does the libmesos version has anything to do with communication. I can use libmesos from previous version and my messages will be correctly transported. Libmesos has no correlation with real Mesos version. Instead of this we should query Mesos master to get it's version.

File: docs/docs/command-line-flags.md (hunk 1 above)
Comment (janisz) [inline docs/docs/command-line-flags.md]: This blog post is partially correct. Mesos Health Checks were introduced as an **experimental** feature in [Mesos 1.1.0](https://github.com/apache/mesos/blob/1.1.0/CHANGELOG#L15-L30)

Comment (janisz) [inline docs/docs/command-line-flags.md]: I `git blame` this file and it looks like it's Jekyll implementation
(escape underscore. Otherwise jekyl will do a <em> around it)[https://github.com/mesosphere/marathon/commit/6462bb992c43bea1866afcd95f44d0c59c673d34]

Comment (janisz) [inline docs/docs/command-line-flags.md]: Fixed

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala (hunk 3 above)
Comment (unterstein) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: hmm .. feature flag should be enough IMO. What do you say @meichstedt ?

Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: tl;dr: fine with me.

Long version:
We don't check master version for other feature flags, so I can't say we must do that here. Yet in general, it would be better to combine feature flags with a minimum master version to make sure that the user gets a meaningful error message when trying to enable a feature that isn't supported by the used mesos version.

Although Mesos tries to be pretty much backwards compatible, I'm sure I have seen invalid combinations of super native library with a newer Mesos. Current master code checks for both libmesos version and master version [here](https://github.com/mesosphere/marathon/blob/23cb098275f0db1a9afda0d59ca9144e99536937/src/main/scala/mesosphere/mesos/LibMesos.scala#L7)

The `MesosMasterMinimumVersion` is checked against after having registered – if the version reported by the frameworkInfo is less, Marathon will throw and commit suicide. Libmesos will be checked during startup (even before registering).

Btw the version as reported by libmesos returned bogus values until [MESOS-6502](https://issues.apache.org/jira/browse/MESOS-6502) was fixed ([here](https://github.com/apache/mesos/commit/cbbc5a46a499eb125eac83af1f21b832f136385a))

**Comment by jdef** [comment]
This is a cool PR! That said, we normally don't backport large features changes like this into a stable branch. We should **proceed with caution** regarding where these changes end up landing.

File: project/build.scala [hunk 4]
```diff
@@ -329,8 +329,8 @@ object Dependency {
     val Chaos = "0.8.7"
     val Guava = "19.0"
     // FIXME (gkleiman): reenable deprecation checks after Mesos 1.0.0-rc2 deprecations are handled
-    val MesosUtils = "1.0.2"
-    val Akka = "2.4.7"
+    val MesosUtils = "1.2.0"
```
Comment (jdef) [inline project/build.scala]: this is not guarded by feature-flag; implications is that future builds of Marathon depend on semantics of newer API, yes?

File: scalastyle-config.xml [hunk 5]
```diff
@@ -1,11 +1,6 @@
 <scalastyle commentFilter="enabled">
  <name>Mesosphere Marathon configuration</name>
  <check level="error" class="org.scalastyle.file.FileTabChecker" enabled="true"/>
- <check level="error" class="org.scalastyle.file.FileLengthChecker" enabled="true">
```
Comment (jdef) [inline scalastyle-config.xml]: why remove? seems unrelated to the nature of this PR

File: src/main/scala/mesosphere/marathon/api/v2/json/AppUpdate.scala [hunk 6]
```diff
@@ -53,7 +53,7 @@ case class AppUpdate(
 
     container: Option[Container] = None,
 
-    healthChecks: Option[Set[HealthCheck]] = None,
+    healthChecks: Option[Set[_ <: HealthCheck]] = None,
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/api/v2/json/AppUpdate.scala]: do we really need a complicated type like this?

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala [hunk 7]
```diff
@@ -72,7 +71,7 @@ case class AppDefinition(
 
   container: Option[Container] = AppDefinition.DefaultContainer,
 
-  healthChecks: Set[HealthCheck] = AppDefinition.DefaultHealthChecks,
+  healthChecks: Set[_ <: HealthCheck] = AppDefinition.DefaultHealthChecks,
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: why the complex type?

File: scalastyle-config.xml (hunk 5 above)
Comment (unterstein) [inline scalastyle-config.xml]: Maybe @gkleiman can give more context here.

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala (hunk 7 above)
Comment (unterstein) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: Maybe @gkleiman can give more context here.

File: src/main/scala/mesosphere/marathon/api/v2/json/AppUpdate.scala (hunk 6 above)
Comment (unterstein) [inline src/main/scala/mesosphere/marathon/api/v2/json/AppUpdate.scala]: Maybe @gkleiman can give more context here.

File: project/build.scala (hunk 4 above)
Comment (janisz) [inline project/build.scala]: Correct. We can downgrade to 1.1 here.

File: src/main/scala/mesosphere/marathon/api/v2/json/AppUpdate.scala (hunk 6 above)
Comment (janisz) [inline src/main/scala/mesosphere/marathon/api/v2/json/AppUpdate.scala]: > [Scala immutable sets are not covariant.](https://github.com/mesosphere/marathon/pull/4210#issuecomment-244425335)

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala (hunk 7 above)
Comment (janisz) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: > [Scala immutable sets are not covariant.](https://github.com/mesosphere/marathon/pull/4210#issuecomment-244425335)

File: docs/docs/command-line-flags.md [hunk 8]
```diff
@@ -47,6 +47,7 @@ The core functionality flags can be also set by environment variable `MARATHON_O
     - "vips" can be used to enable the networking VIP integration UI.
     - "task\_killing" can be used to enable the TASK\_KILLING state in Mesos (0.28 or later)
     - "external\_volumes" can be used if the cluster is configured to use external volumes.
+    - "mesos\_healthchecks" can be used if ti enable the Mesos HealthChecks (Mesos 1.1 or later).
```
Comment (janisz) [inline docs/docs/command-line-flags.md]: typo: `if ti` → `to`

File: scalastyle-config.xml (hunk 5 above)
Comment (janisz) [inline scalastyle-config.xml]: I agree with @jdef this shouldn't be touched. Should I revert it or keep it consistent with 1.4?

**Comment by janisz** [comment]
> it introduces a dependency on a newer Mesos API (with breaking changes): for example, custom executors have new proto requirements for the 1.2 API.

@jdef What breaking changes do you refer to? I haven't noticed any breaking changes? Noting is mentioned in [Mesos changelog](https://github.com/apache/mesos/blob/1.2.0/CHANGELOG) too? Can you elaborate.

**Comment by jdef** [comment]
I built a toy framework in mesos-go that uses a custom executor. I didn't
need to make any proto-building changes when upgrading from v1.0.x to
v1.1.x mesos protos. Upon upgrading to mesos v1.2.x protos the logic that
built a custom executor needs to set the executor's type to CUSTOM
otherwise mesos rejects it. Maybe this is an under-documented feature of
Mesos 1.2.x?

**Comment by janisz** [comment]
We are using custom executor with Marathon and we didn't need to change anything while upgrading Marathon (1.3 then 1.4.2 then 1.3 with HC). [Marathon still doesn't set executors properly](https://github.com/mesosphere/marathon/blame/48bfffdbb2920732ba726ce19ade6deb00121794/src/main/scala/mesosphere/mesos/TaskBuilder.scala#L62-L77) ([see also MARATHON-4210](https://jira.mesosphere.com/browse/MARATHON-4210)). Executor code lives untouched from ages and [nobody knows why/how it's working](https://jira.mesosphere.com/browse/MARATHON-7191?focusedCommentId=100100&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-100100).

**Comment by jdef** [comment]
oh boy. maybe there's a difference between the mesos v0 and v1 API
implementation?!

**Comment by timcharper** [comment]
@janisz do you still feel like you need this change?

I'd love to just merge it as is, it is really great work! However, I got pushback on the idea of supporting a Mesos 1.1.0 feature in a version of Marathon (1.3.x) that targets Mesos 1.0.0. I thought of pushing a custom build of this change, but, this seems like a less-than-optimal road, too. Perhaps you create a release and upload the artifact to your fork, so others can find it?

With regard to this comment:

> Moving all Marathon applications from Marathon healthcecks to Mesos Healthcecks can't be done at once. 

Is it not possible to move to Mesos health checks after upgrading to Marathon 1.4.x?

Hopefully we have fixed the blocked issues to upgrade to Marathon 1.4.x. We changed a lot of internals and, unfortunately, it has been a bit of a bumpy ride.

**Comment by timcharper** [comment]
Closing :(

We couldn't accept this for 1.3.x because it breaks our version compatibility matrix. The version of DCOS that ships Marathon 1.3.x also ships a version of Mesos that does not support these health checks.

Sorry about all this. I would like to find a way to improve our collaboration process where such changes can be discussed first (and consensus achieved) before changes are made. I wish Marathon were a little nimble open source project that we could just run with and change as it made sense. The reality is a pretty big picture around it, so, we need to talk about these types of changes and make sure we understand the impact. I failed to do so prior to giving my accept for the PR (and, have failed to do so, in the past, on some other ones, because I really hate saying no to PRs). This applies to core Marathon committers, too.

I still hate saying no to PRs. I'd much prefer to redirect effort, earlier, so we can say yes, later.

**Comment by janisz** [comment]
@timcharper No problem. I was aware that there is small chance it will be merged (I should have written it in the first comment). I was thinking about closing this PR earlier but maybe there was someone who needed it. This PR was created to show the community there is a chance to get Mesos HealthChecks without Marathon 1.4 that was "unstable" at that time. We are happily running Marathon 1.3 with this patch and it saves from performance issues described [here](https://mesosphere.com/blog/introducing-mesos-native-health-checks-apache-mesos-part-1/). I hope we soon update to 1.4.

> I still hate saying no to PRs. I'd much prefer to redirect effort, earlier, so we can say yes, later.

Agree, I'll prepare some proposal next time before submitting a patch.
